
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://fulopjoz.github.io/pi-avg/alignments/global_local/">
      
      
        <link rel="prev" href="../edit_distance/">
      
      
        <link rel="next" href="../affine_gap/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.7">
    
    
      
        <title>Global & Local Alignment - PI-AVG: Computational Genomics Algorithms</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CJetBrains+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"JetBrains Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_markdown_exec_pyodide.css">
    
      <link rel="stylesheet" href="../../assets/css/custom.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#global-local-and-overlap-alignment" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="PI-AVG: Computational Genomics Algorithms" class="md-header__button md-logo" aria-label="PI-AVG: Computational Genomics Algorithms" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PI-AVG: Computational Genomics Algorithms
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Global & Local Alignment
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-2.69L23.31 12 20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/fulopjoz/pi-avg" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    PI-AVG Computational Genomics Algorithms
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Sequence Alignment

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../algorithms/" class="md-tabs__link">
          
  
  Algorithms

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../data_structures/" class="md-tabs__link">
          
  
  Data Structures

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../case_studies/" class="md-tabs__link">
          
  
  Case Studies

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../code_examples/" class="md-tabs__link">
          
  
  Code Examples

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="PI-AVG: Computational Genomics Algorithms" class="md-nav__button md-logo" aria-label="PI-AVG: Computational Genomics Algorithms" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    PI-AVG: Computational Genomics Algorithms
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/fulopjoz/pi-avg" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    PI-AVG Computational Genomics Algorithms
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Sequence Alignment
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Sequence Alignment
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Basic Concepts
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../edit_distance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Edit Distance
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Global & Local Alignment
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Global & Local Alignment
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#global-alignment-needleman-wunsch" class="md-nav__link">
    <span class="md-ellipsis">
      Global Alignment (Needleman-Wunsch)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Global Alignment (Needleman-Wunsch)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#needleman-wunsch-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      Needleman-Wunsch Algorithm
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Needleman-Wunsch Algorithm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#algorithm-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Steps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#local-alignment-smith-waterman" class="md-nav__link">
    <span class="md-ellipsis">
      Local Alignment (Smith-Waterman)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Local Alignment (Smith-Waterman)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#smith-waterman-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      Smith-Waterman Algorithm
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Smith-Waterman Algorithm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#algorithm-steps_1" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Steps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example_1" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#overlap-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Overlap Alignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overlap Alignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#algorithm-approach" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Approach
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#choosing-the-right-alignment-approach" class="md-nav__link">
    <span class="md-ellipsis">
      Choosing the Right Alignment Approach
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#similarity-scoring-in-biological-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      Similarity Scoring in Biological Sequences
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Similarity Scoring in Biological Sequences">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dnarna-scoring" class="md-nav__link">
    <span class="md-ellipsis">
      DNA/RNA Scoring
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#protein-scoring" class="md-nav__link">
    <span class="md-ellipsis">
      Protein Scoring
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../affine_gap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Affine Gap Penalties
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../multiple_sequence/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Multiple Sequence Alignment
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Algorithms
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Algorithms
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../algorithms/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Data Structures
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Data Structures
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data_structures/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Case Studies
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Case Studies
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../case_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Examples
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Code Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../code_examples/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#global-alignment-needleman-wunsch" class="md-nav__link">
    <span class="md-ellipsis">
      Global Alignment (Needleman-Wunsch)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Global Alignment (Needleman-Wunsch)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#needleman-wunsch-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      Needleman-Wunsch Algorithm
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Needleman-Wunsch Algorithm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#algorithm-steps" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Steps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#local-alignment-smith-waterman" class="md-nav__link">
    <span class="md-ellipsis">
      Local Alignment (Smith-Waterman)
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Local Alignment (Smith-Waterman)">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#smith-waterman-algorithm" class="md-nav__link">
    <span class="md-ellipsis">
      Smith-Waterman Algorithm
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Smith-Waterman Algorithm">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#algorithm-steps_1" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Steps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example_1" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#overlap-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Overlap Alignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overlap Alignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#algorithm-approach" class="md-nav__link">
    <span class="md-ellipsis">
      Algorithm Approach
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#choosing-the-right-alignment-approach" class="md-nav__link">
    <span class="md-ellipsis">
      Choosing the Right Alignment Approach
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#similarity-scoring-in-biological-sequences" class="md-nav__link">
    <span class="md-ellipsis">
      Similarity Scoring in Biological Sequences
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Similarity Scoring in Biological Sequences">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dnarna-scoring" class="md-nav__link">
    <span class="md-ellipsis">
      DNA/RNA Scoring
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#protein-scoring" class="md-nav__link">
    <span class="md-ellipsis">
      Protein Scoring
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  
    
      
    
    <a href="https://github.com/fulopjoz/pi-avg/raw/main/docs/alignments/global_local.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0 8a5 5 0 0 1-5-5 5 5 0 0 1 5-5 5 5 0 0 1 5 5 5 5 0 0 1-5 5m0-12.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5"/></svg>
    </a>
  


<h1 id="global-local-and-overlap-alignment">Global, Local, and Overlap Alignment<a class="headerlink" href="#global-local-and-overlap-alignment" title="Permanent link">&para;</a></h1>
<p>While edit distance focuses on the minimum number of operations to transform one sequence into another, sequence alignment algorithms typically use scoring schemes to find the optimal alignment between sequences. There are three main types of alignment approaches: global, local, and overlap alignment.</p>
<h2 id="global-alignment-needleman-wunsch">Global Alignment (Needleman-Wunsch)<a class="headerlink" href="#global-alignment-needleman-wunsch" title="Permanent link">&para;</a></h2>
<p><strong>Global alignment</strong> attempts to align two sequences from start to end. This approach is most useful when:
- The sequences are of similar length
- The sequences are expected to be similar throughout their entire length
- You want to find the best overall alignment</p>
<h3 id="needleman-wunsch-algorithm">Needleman-Wunsch Algorithm<a class="headerlink" href="#needleman-wunsch-algorithm" title="Permanent link">&para;</a></h3>
<p>The <strong>Needleman-Wunsch algorithm</strong> is the standard dynamic programming approach for global alignment. It uses a scoring scheme rather than a cost function:</p>
<ul>
<li>Match: positive score (e.g., +1)</li>
<li>Mismatch: negative score (e.g., -1)</li>
<li>Gap: negative score (e.g., -2)</li>
</ul>
<h4 id="algorithm-steps">Algorithm Steps<a class="headerlink" href="#algorithm-steps" title="Permanent link">&para;</a></h4>
<ol>
<li><strong>Initialization</strong>: Create a scoring matrix <span class="arithmatex">\(S\)</span> of size <span class="arithmatex">\((n+1) \times (m+1)\)</span></li>
<li><span class="arithmatex">\(S(0,0) = 0\)</span></li>
<li><span class="arithmatex">\(S(i,0) = -i \times \text{gap\_penalty}\)</span> for all <span class="arithmatex">\(i &gt; 0\)</span></li>
<li>
<p><span class="arithmatex">\(S(0,j) = -j \times \text{gap\_penalty}\)</span> for all <span class="arithmatex">\(j &gt; 0\)</span></p>
</li>
<li>
<p><strong>Fill the matrix</strong>: For each cell <span class="arithmatex">\(S(i,j)\)</span>, compute:
   $$
   S(i,j) = \max \begin{cases}
   S(i-1,j-1) + s(A[i], B[j]) &amp; \text{(match/mismatch)} \
   S(i-1,j) - \text{gap_penalty} &amp; \text{(deletion)} \
   S(i,j-1) - \text{gap_penalty} &amp; \text{(insertion)}
   \end{cases}
   $$
   Where <span class="arithmatex">\(s(A[i], B[j])\)</span> is the score for aligning characters <span class="arithmatex">\(A[i]\)</span> and <span class="arithmatex">\(B[j]\)</span>.</p>
</li>
<li>
<p><strong>Traceback</strong>: Start from <span class="arithmatex">\(S(n,m)\)</span> and follow the path that led to each cell's score until reaching <span class="arithmatex">\(S(0,0)\)</span>.</p>
</li>
</ol>
<h4 id="example">Example<a class="headerlink" href="#example" title="Permanent link">&para;</a></h4>
<p>Consider aligning sequences <span class="arithmatex">\(A = \text{"ACGT"}\)</span> and <span class="arithmatex">\(B = \text{"ATGC"}\)</span> with:
- Match score: +1
- Mismatch score: -1
- Gap penalty: -2</p>
<p>The scoring matrix would be:</p>
<div class="highlight"><pre><span></span><code>    |    | A  | C  | G  | T
----|----|----|----|----|----
    |  0 | -2 | -4 | -6 | -8
----|----|----|----|----|----
  A | -2 |  1 | -1 | -3 | -5
----|----|----|----|----|----
  T | -4 | -1 |  0 | -2 | -2
----|----|----|----|----|----
  G | -6 | -3 | -2 |  1 | -1
----|----|----|----|----|----
  C | -8 | -5 | -1 | -1 |  0
</code></pre></div>
<p>The optimal global alignment has a score of 0 and might be:</p>
<div class="highlight"><pre><span></span><code>A C G T
A T G C
</code></pre></div>
<h2 id="local-alignment-smith-waterman">Local Alignment (Smith-Waterman)<a class="headerlink" href="#local-alignment-smith-waterman" title="Permanent link">&para;</a></h2>
<p><strong>Local alignment</strong> aims to find the most similar <strong>substrings</strong> within two sequences. This approach is most useful when:
- The sequences may only share regions of similarity
- You want to identify conserved domains or motifs
- The sequences differ significantly in length or content</p>
<h3 id="smith-waterman-algorithm">Smith-Waterman Algorithm<a class="headerlink" href="#smith-waterman-algorithm" title="Permanent link">&para;</a></h3>
<p>The <strong>Smith-Waterman algorithm</strong> modifies the Needleman-Wunsch approach to focus on local regions of similarity:</p>
<h4 id="algorithm-steps_1">Algorithm Steps<a class="headerlink" href="#algorithm-steps_1" title="Permanent link">&para;</a></h4>
<ol>
<li><strong>Initialization</strong>: Create a scoring matrix <span class="arithmatex">\(L\)</span> of size <span class="arithmatex">\((n+1) \times (m+1)\)</span></li>
<li><span class="arithmatex">\(L(i,0) = 0\)</span> for all <span class="arithmatex">\(i\)</span></li>
<li>
<p><span class="arithmatex">\(L(0,j) = 0\)</span> for all <span class="arithmatex">\(j\)</span></p>
</li>
<li>
<p><strong>Fill the matrix</strong>: For each cell <span class="arithmatex">\(L(i,j)\)</span>, compute:
   $$
   L(i,j) = \max \begin{cases}
   0 &amp; \text{(start new alignment)} \
   L(i-1,j-1) + s(A[i], B[j]) &amp; \text{(match/mismatch)} \
   L(i-1,j) - \text{gap_penalty} &amp; \text{(deletion)} \
   L(i,j-1) - \text{gap_penalty} &amp; \text{(insertion)}
   \end{cases}
   $$</p>
</li>
<li>
<p><strong>Traceback</strong>: Start from the cell with the <strong>highest score</strong> in the entire matrix and follow the path until reaching a cell with score 0.</p>
</li>
</ol>
<p>The key difference from global alignment is that:
- Scores are never allowed to go below zero (we reset to zero instead)
- We start the traceback from the highest-scoring cell (not necessarily the bottom-right)
- The traceback ends when we reach a zero (not necessarily at the top-left)</p>
<h4 id="example_1">Example<a class="headerlink" href="#example_1" title="Permanent link">&para;</a></h4>
<p>Using the same sequences and scoring scheme as before, the local alignment matrix might look like:</p>
<div class="highlight"><pre><span></span><code>    |    | A  | C  | G  | T
----|----|----|----|----|----
    |  0 |  0 |  0 |  0 |  0
----|----|----|----|----|----
  A |  0 |  1 |  0 |  0 |  0
----|----|----|----|----|----
  T |  0 |  0 |  0 |  0 |  1
----|----|----|----|----|----
  G |  0 |  0 |  0 |  1 |  0
----|----|----|----|----|----
  C |  0 |  0 |  1 |  0 |  0
</code></pre></div>
<p>The optimal local alignment might be just a single matching character or a longer substring, depending on the sequences and scoring parameters.</p>
<h2 id="overlap-alignment">Overlap Alignment<a class="headerlink" href="#overlap-alignment" title="Permanent link">&para;</a></h2>
<p><strong>Overlap alignment</strong> (or semi-global alignment) is used when we want to find the best overlap between the end of one sequence and the beginning of another. This is particularly useful in:
- Genome assembly, where we need to merge overlapping sequence fragments
- Identifying potential fusion points between sequences</p>
<h3 id="algorithm-approach">Algorithm Approach<a class="headerlink" href="#algorithm-approach" title="Permanent link">&para;</a></h3>
<p>Overlap alignment modifies the global alignment algorithm by:
- Not penalizing gaps at the beginning of one sequence and/or the end of the other
- This is implemented by modifying the initialization and/or the final score selection</p>
<p>For example, to find the best suffix of <span class="arithmatex">\(A\)</span> that matches a prefix of <span class="arithmatex">\(B\)</span>:
1. Initialize the first row normally: <span class="arithmatex">\(S(0,j) = -j \times \text{gap\_penalty}\)</span>
2. Initialize the first column with zeros: <span class="arithmatex">\(S(i,0) = 0\)</span> (no penalty for gaps at the start of <span class="arithmatex">\(B\)</span>)
3. Fill the matrix as in global alignment
4. For the final score, find the maximum value in the last row (not necessarily <span class="arithmatex">\(S(n,m)\)</span>)</p>
<h2 id="choosing-the-right-alignment-approach">Choosing the Right Alignment Approach<a class="headerlink" href="#choosing-the-right-alignment-approach" title="Permanent link">&para;</a></h2>
<p>The choice between global, local, and overlap alignment depends on the biological context:</p>
<table>
<thead>
<tr>
<th>Alignment Type</th>
<th>When to Use</th>
</tr>
</thead>
<tbody>
<tr>
<td>Global</td>
<td>When sequences are similar throughout (e.g., homologous genes)</td>
</tr>
<tr>
<td>Local</td>
<td>When looking for conserved domains or motifs in otherwise dissimilar sequences</td>
</tr>
<tr>
<td>Overlap</td>
<td>When assembling fragments or finding potential fusion points</td>
</tr>
</tbody>
</table>
<h2 id="similarity-scoring-in-biological-sequences">Similarity Scoring in Biological Sequences<a class="headerlink" href="#similarity-scoring-in-biological-sequences" title="Permanent link">&para;</a></h2>
<p>In biological sequence alignment, the scoring scheme is crucial and often reflects evolutionary or biochemical properties:</p>
<h3 id="dnarna-scoring">DNA/RNA Scoring<a class="headerlink" href="#dnarna-scoring" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>Match</strong>: Typically +1 or +5</li>
<li><strong>Mismatch</strong>: Often -1 to -4, but may vary based on transition vs. transversion</li>
<li><strong>Gap</strong>: Usually a larger penalty, e.g., -2 to -10</li>
</ul>
<h3 id="protein-scoring">Protein Scoring<a class="headerlink" href="#protein-scoring" title="Permanent link">&para;</a></h3>
<p>For proteins, substitution matrices like PAM (Point Accepted Mutation) or BLOSUM (BLOcks SUbstitution Matrix) are used:
- These matrices assign different scores to different amino acid substitutions based on observed frequencies in related proteins
- For example, substituting similar amino acids (e.g., leucine for isoleucine) gets a higher score than dissimilar ones</p>
<p>In the next section, we'll explore affine gap penalties, which provide a more realistic model for biological insertions and deletions.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "navigation.top", "search.highlight", "content.action.view", "content.code.copy"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../../assets/_markdown_exec_pyodide.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
    
  </body>
</html>