
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://fulopjoz.github.io/pi-avg/alignments/multiple_sequence/">
      
      
        <link rel="prev" href="../affine_gap/">
      
      
        <link rel="next" href="../../algorithms/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.7">
    
    
      
        <title>Multiple Sequence Alignment - PI-AVG: Computational Genomics Algorithms</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CJetBrains+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"JetBrains Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../assets/_markdown_exec_pyodide.css">
    
      <link rel="stylesheet" href="../../assets/css/custom.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#multiple-sequence-alignment" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="PI-AVG: Computational Genomics Algorithms" class="md-header__button md-logo" aria-label="PI-AVG: Computational Genomics Algorithms" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            PI-AVG: Computational Genomics Algorithms
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Multiple Sequence Alignment
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a4 4 0 0 0-4 4 4 4 0 0 0 4 4 4 4 0 0 0 4-4 4 4 0 0 0-4-4m0 10a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-9.31V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20v-4.69L23.31 12z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 18a6 6 0 0 1-6-6 6 6 0 0 1 6-6 6 6 0 0 1 6 6 6 6 0 0 1-6 6m8-2.69L23.31 12 20 8.69V4h-4.69L12 .69 8.69 4H4v4.69L.69 12 4 15.31V20h4.69L12 23.31 15.31 20H20z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/fulopjoz/pi-avg" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    PI-AVG Computational Genomics Algorithms
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../.." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../" class="md-tabs__link">
          
  
  Sequence Alignment

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../algorithms/" class="md-tabs__link">
          
  
  Algorithms

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../data_structures/" class="md-tabs__link">
          
  
  Data Structures

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../case_studies/" class="md-tabs__link">
          
  
  Case Studies

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../code_examples/" class="md-tabs__link">
          
  
  Code Examples

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="PI-AVG: Computational Genomics Algorithms" class="md-nav__button md-logo" aria-label="PI-AVG: Computational Genomics Algorithms" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    PI-AVG: Computational Genomics Algorithms
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/fulopjoz/pi-avg" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6m-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3m44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9M244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8M97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1m-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7m32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1m-11.4-14.7c-1.6 1-1.6 3.6 0 5.9s4.3 3.3 5.6 2.3c1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2"/></svg>
  </div>
  <div class="md-source__repository">
    PI-AVG Computational Genomics Algorithms
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Sequence Alignment
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Sequence Alignment
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../basic/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Basic Concepts
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../edit_distance/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Edit Distance
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../global_local/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Global & Local Alignment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../affine_gap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Affine Gap Penalties
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Multiple Sequence Alignment
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Multiple Sequence Alignment
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#definition-and-motivation" class="md-nav__link">
    <span class="md-ellipsis">
      Definition and Motivation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#applications" class="md-nav__link">
    <span class="md-ellipsis">
      Applications
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pairwise-vs-multiple-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Pairwise vs. Multiple Alignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scoring-schemes" class="md-nav__link">
    <span class="md-ellipsis">
      Scoring Schemes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Scoring Schemes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sum-of-pairs-sp-score" class="md-nav__link">
    <span class="md-ellipsis">
      Sum-of-Pairs (SP) Score
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tree-based-evolutionary-scoring" class="md-nav__link">
    <span class="md-ellipsis">
      Tree-based / Evolutionary Scoring
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gap-penalties" class="md-nav__link">
    <span class="md-ellipsis">
      Gap Penalties
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#practical-approaches" class="md-nav__link">
    <span class="md-ellipsis">
      Practical Approaches
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Practical Approaches">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#progressive-guide-tree-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Progressive (Guide-Tree) Alignment
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-heuristics" class="md-nav__link">
    <span class="md-ellipsis">
      Other Heuristics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#special-cases-in-sequence-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Special Cases in Sequence Alignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Special Cases in Sequence Alignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dna-protein-alignment-and-codons" class="md-nav__link">
    <span class="md-ellipsis">
      DNA-Protein Alignment and Codons
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#handling-introns-and-large-gaps" class="md-nav__link">
    <span class="md-ellipsis">
      Handling Introns and Large Gaps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#complexity-and-speed-ups" class="md-nav__link">
    <span class="md-ellipsis">
      Complexity and Speed-Ups
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Complexity and Speed-Ups">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#multiple-alignment-complexity" class="md-nav__link">
    <span class="md-ellipsis">
      Multiple Alignment Complexity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pairwise-alignment-complexity" class="md-nav__link">
    <span class="md-ellipsis">
      Pairwise Alignment Complexity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#advanced-techniques" class="md-nav__link">
    <span class="md-ellipsis">
      Advanced Techniques
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#worked-example-protein-dna-with-introns" class="md-nav__link">
    <span class="md-ellipsis">
      Worked Example: Protein-DNA with Introns
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#practical-applications" class="md-nav__link">
    <span class="md-ellipsis">
      Practical Applications
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#common-questions-and-answers" class="md-nav__link">
    <span class="md-ellipsis">
      Common Questions and Answers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#practice-questions" class="md-nav__link">
    <span class="md-ellipsis">
      Practice Questions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-reading" class="md-nav__link">
    <span class="md-ellipsis">
      Additional Reading
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Algorithms
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Algorithms
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../algorithms/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Data Structures
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Data Structures
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../data_structures/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Case Studies
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Case Studies
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../case_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" >
        
          
          <label class="md-nav__link" for="__nav_6" id="__nav_6_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Code Examples
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Code Examples
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../code_examples/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#definition-and-motivation" class="md-nav__link">
    <span class="md-ellipsis">
      Definition and Motivation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#applications" class="md-nav__link">
    <span class="md-ellipsis">
      Applications
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#pairwise-vs-multiple-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Pairwise vs. Multiple Alignment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scoring-schemes" class="md-nav__link">
    <span class="md-ellipsis">
      Scoring Schemes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Scoring Schemes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#sum-of-pairs-sp-score" class="md-nav__link">
    <span class="md-ellipsis">
      Sum-of-Pairs (SP) Score
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#tree-based-evolutionary-scoring" class="md-nav__link">
    <span class="md-ellipsis">
      Tree-based / Evolutionary Scoring
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gap-penalties" class="md-nav__link">
    <span class="md-ellipsis">
      Gap Penalties
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#practical-approaches" class="md-nav__link">
    <span class="md-ellipsis">
      Practical Approaches
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Practical Approaches">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#progressive-guide-tree-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Progressive (Guide-Tree) Alignment
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#other-heuristics" class="md-nav__link">
    <span class="md-ellipsis">
      Other Heuristics
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#special-cases-in-sequence-alignment" class="md-nav__link">
    <span class="md-ellipsis">
      Special Cases in Sequence Alignment
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Special Cases in Sequence Alignment">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dna-protein-alignment-and-codons" class="md-nav__link">
    <span class="md-ellipsis">
      DNA-Protein Alignment and Codons
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#handling-introns-and-large-gaps" class="md-nav__link">
    <span class="md-ellipsis">
      Handling Introns and Large Gaps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#complexity-and-speed-ups" class="md-nav__link">
    <span class="md-ellipsis">
      Complexity and Speed-Ups
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Complexity and Speed-Ups">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#multiple-alignment-complexity" class="md-nav__link">
    <span class="md-ellipsis">
      Multiple Alignment Complexity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#pairwise-alignment-complexity" class="md-nav__link">
    <span class="md-ellipsis">
      Pairwise Alignment Complexity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#advanced-techniques" class="md-nav__link">
    <span class="md-ellipsis">
      Advanced Techniques
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#worked-example-protein-dna-with-introns" class="md-nav__link">
    <span class="md-ellipsis">
      Worked Example: Protein-DNA with Introns
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#practical-applications" class="md-nav__link">
    <span class="md-ellipsis">
      Practical Applications
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#common-questions-and-answers" class="md-nav__link">
    <span class="md-ellipsis">
      Common Questions and Answers
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#practice-questions" class="md-nav__link">
    <span class="md-ellipsis">
      Practice Questions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#additional-reading" class="md-nav__link">
    <span class="md-ellipsis">
      Additional Reading
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  
    
      
    
    <a href="https://github.com/fulopjoz/pi-avg/raw/main/docs/alignments/multiple_sequence.md" title="View source of this page" class="md-content__button md-icon">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0 8a5 5 0 0 1-5-5 5 5 0 0 1 5-5 5 5 0 0 1 5 5 5 5 0 0 1-5 5m0-12.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5"/></svg>
    </a>
  


<h1 id="multiple-sequence-alignment">Multiple Sequence Alignment<a class="headerlink" href="#multiple-sequence-alignment" title="Permanent link">&para;</a></h1>
<h2 id="definition-and-motivation">Definition and Motivation<a class="headerlink" href="#definition-and-motivation" title="Permanent link">&para;</a></h2>
<p>Multiple Sequence Alignment (MSA) involves aligning three or more biological sequences (DNA, RNA, or protein) so that homologous positions line up in the same columns, reflecting biological or evolutionary relationships.</p>
<p>For sequences <span class="arithmatex">\(X_1, X_2, \dots, X_D\)</span>, an MSA can be viewed as a matrix with <span class="arithmatex">\(D\)</span> rows, each row being <span class="arithmatex">\(X_i\)</span> padded with gaps (<span class="arithmatex">\(-\)</span>) so that all rows have equal length <span class="arithmatex">\(h\)</span>. Formally:</p>
<ul>
<li>Let <span class="arithmatex">\(X_i = X_i[1..n_i]\)</span></li>
<li>An alignment is <span class="arithmatex">\(\bigl(U_1, U_2, \dots, U_D \bigr)\)</span></li>
<li>Where each <span class="arithmatex">\(U_i\)</span> is length <span class="arithmatex">\(h\)</span> and is <span class="arithmatex">\(X_i\)</span> with <span class="arithmatex">\((h - n_i)\)</span> gaps inserted</li>
</ul>
<h2 id="applications">Applications<a class="headerlink" href="#applications" title="Permanent link">&para;</a></h2>
<p>Multiple sequence alignment has numerous applications in computational genomics:</p>
<ol>
<li><strong>Identifying conserved motifs</strong> across related species or genes</li>
<li><strong>Inferring evolutionary history</strong> (e.g., building phylogenetic trees)</li>
<li><strong>Improving sensitivity in database searches</strong> (profiles, HMMs)</li>
<li><strong>Aligning genomic sequences</strong> that may contain introns, exons, and large gaps</li>
</ol>
<h2 id="pairwise-vs-multiple-alignment">Pairwise vs. Multiple Alignment<a class="headerlink" href="#pairwise-vs-multiple-alignment" title="Permanent link">&para;</a></h2>
<p>While pairwise alignment (comparing two sequences) can be solved efficiently using dynamic programming in polynomial time (e.g., <span class="arithmatex">\(O(nm)\)</span> for sequences of length <span class="arithmatex">\(n\)</span> and <span class="arithmatex">\(m\)</span>), multiple sequence alignment presents greater computational challenges:</p>
<ul>
<li><strong>Pairwise alignment</strong>: Generally solvable in polynomial time</li>
<li><strong>Multiple sequence alignment</strong>: NP-hard in the general case when trying to optimize a standard objective function</li>
</ul>
<p>This computational complexity necessitates the use of heuristic approaches for aligning multiple sequences.</p>
<h2 id="scoring-schemes">Scoring Schemes<a class="headerlink" href="#scoring-schemes" title="Permanent link">&para;</a></h2>
<h3 id="sum-of-pairs-sp-score">Sum-of-Pairs (SP) Score<a class="headerlink" href="#sum-of-pairs-sp-score" title="Permanent link">&para;</a></h3>
<p>One of the most common scoring schemes for MSA is the Sum-of-Pairs (SP) score:</p>
<div class="arithmatex">\[
\text{Score}(\text{MSA}) \;=\;
\sum_{1 \le p &lt; q \le D} \bigl(\text{score of pairwise alignment of } U_p, U_q\bigr)
\]</div>
<p>This approach sums the scores of all possible pairwise alignments within the MSA.</p>
<h3 id="tree-based-evolutionary-scoring">Tree-based / Evolutionary Scoring<a class="headerlink" href="#tree-based-evolutionary-scoring" title="Permanent link">&para;</a></h3>
<p>Instead of just summing pairs, one can weight edges according to a phylogenetic tree, which may better reflect evolutionary relationships.</p>
<h3 id="gap-penalties">Gap Penalties<a class="headerlink" href="#gap-penalties" title="Permanent link">&para;</a></h3>
<p>Gaps may be penalized differently if they extend (affine gap costs) or if they appear in multiple rows. The choice of gap penalty scheme can significantly impact the resulting alignment.</p>
<h2 id="practical-approaches">Practical Approaches<a class="headerlink" href="#practical-approaches" title="Permanent link">&para;</a></h2>
<p>Due to the NP-hardness of finding an optimal MSA, several heuristic approaches have been developed:</p>
<h3 id="progressive-guide-tree-alignment">Progressive (Guide-Tree) Alignment<a class="headerlink" href="#progressive-guide-tree-alignment" title="Permanent link">&para;</a></h3>
<p>This is one of the most widely used approaches:</p>
<ol>
<li>First, construct a pairwise distance matrix for all sequences</li>
<li>Build a guide tree (e.g., neighbor-joining or UPGMA)</li>
<li>Align the two most similar sequences to get a profile</li>
<li>Align the next sequence (or profile) to the existing alignment, continuing until all are included</li>
</ol>
<p><strong>Profiles</strong>: Each column is summarized by frequencies of A, C, G, T (for DNA) or 20 amino acids (for proteins). When aligning two profiles, each "symbol" in the dynamic programming matrix is a distribution of characters rather than a single character.</p>
<h3 id="other-heuristics">Other Heuristics<a class="headerlink" href="#other-heuristics" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>Iterative refinement</strong>: Start with a progressive alignment, then realign subgroups to escape local maxima</li>
<li><strong>Consistency-based methods</strong>: Like T-Coffee, using pairwise alignment libraries</li>
</ul>
<h2 id="special-cases-in-sequence-alignment">Special Cases in Sequence Alignment<a class="headerlink" href="#special-cases-in-sequence-alignment" title="Permanent link">&para;</a></h2>
<h3 id="dna-protein-alignment-and-codons">DNA-Protein Alignment and Codons<a class="headerlink" href="#dna-protein-alignment-and-codons" title="Permanent link">&para;</a></h3>
<p>When aligning DNA to protein sequences, special considerations are needed:</p>
<ul>
<li><strong>Codon-based alignment</strong>: Each amino acid (in a protein) is typically encoded by a triplet of bases (codon) in DNA</li>
<li><strong>Possible reading frames</strong>: DNA can be read in frames 0, 1, or 2 (or all six frames if reverse complement is considered)</li>
<li><strong>Implementation</strong>: This often requires a "3-to-1" stepping in the dynamic programming algorithm</li>
</ul>
<p>In a codon-based alignment approach:
- For the DNA index: increment in steps of 3 (because one codon = three nucleotides)
- For the protein index: increment in steps of 1 (each index is one amino acid)</p>
<p>A conceptual Needleman-Wunsch style recurrence adapted for codon-based alignment might look like:</p>
<div class="arithmatex">\[
\begin{aligned}
\text{score}[i,j] = \max\Bigl\{&amp;\;\text{score}[\,i-1,\,j-3\,] + \text{subst}\bigl(\text{DNA}[j-2,j-1,j],\,\text{Protein}[i]\bigr),\\
&amp;\;\text{score}[\,i,\,j-1\,] - \delta_{\text{gap}},\\
&amp;\;\text{score}[\,i-1,\,j\,] - \delta_{\text{gap}} \Bigr\}.
\end{aligned}
\]</div>
<p>Where:
- <span class="arithmatex">\(i\)</span> indexes the protein from 1..<span class="arithmatex">\(m\)</span> (where <span class="arithmatex">\(m\)</span> is amino acid length)
- <span class="arithmatex">\(j\)</span> indexes the DNA from 1..<span class="arithmatex">\(3n\)</span> (where <span class="arithmatex">\(n\)</span> is the number of codons)
- The first line represents the codon→amino-acid match or mismatch
- The other lines allow insertion/deletion (effectively handling frameshifts or partial alignments)</p>
<h3 id="handling-introns-and-large-gaps">Handling Introns and Large Gaps<a class="headerlink" href="#handling-introns-and-large-gaps" title="Permanent link">&para;</a></h3>
<p>In eukaryotic genes, large introns disrupt the coding exons, presenting challenges for alignment:</p>
<ul>
<li><strong>Intron-aware alignment</strong>: Because introns can be long and variable in length, standard gap penalties may underestimate or mis-score them</li>
<li><strong>Spliced alignment</strong>: When aligning cDNA to genomic DNA, introns appear as large gaps in the genomic sequence</li>
<li><strong>Specialized approaches</strong>: Tools often use specialized spliced alignment approaches for gene alignment</li>
</ul>
<p>Introns can be thousands—or in some cases, millions—of bases long, creating significant challenges:
- Pure dynamic programming with large gap penalties can become computationally expensive
- Bounds on intron length may be needed (average intron size is around ~1,500 bases, but maximum introns can exceed 1 million bases)
- Spliced alignment tools try to find exons that match the protein's coding sequence and allow skipping entire introns</p>
<h2 id="complexity-and-speed-ups">Complexity and Speed-Ups<a class="headerlink" href="#complexity-and-speed-ups" title="Permanent link">&para;</a></h2>
<h3 id="multiple-alignment-complexity">Multiple Alignment Complexity<a class="headerlink" href="#multiple-alignment-complexity" title="Permanent link">&para;</a></h3>
<ul>
<li>Exact dynamic programming for <span class="arithmatex">\(D\)</span> sequences of average length <span class="arithmatex">\(N\)</span>: <span class="arithmatex">\(O(N^D)\)</span> time</li>
<li>This exponential complexity makes exact solutions impractical for more than a few sequences</li>
</ul>
<h3 id="pairwise-alignment-complexity">Pairwise Alignment Complexity<a class="headerlink" href="#pairwise-alignment-complexity" title="Permanent link">&para;</a></h3>
<ul>
<li><span class="arithmatex">\(O(nm)\)</span> time, <span class="arithmatex">\(O(\min\{n,m\})\)</span> space for global/local alignment</li>
<li>For DNA-protein alignment with long genomic DNA (millions of nucleotides) vs. a protein of length ~300, the computation can be extensive</li>
</ul>
<h3 id="advanced-techniques">Advanced Techniques<a class="headerlink" href="#advanced-techniques" title="Permanent link">&para;</a></h3>
<ul>
<li><strong>Affine Gap Costs</strong>: Extend the dynamic programming with extra "tracks" for gap opening vs. gap extension</li>
<li><strong>Bit-Parallel Methods</strong>: For approximate string matching with small alphabets (like Myers Bit-Parallel algorithm)</li>
<li><strong>Heuristic Pruning</strong>: Limiting the search space based on biological constraints</li>
<li><strong>Index structures</strong>: Using suffix arrays or BWT-based structures for short local alignments, then chaining them together</li>
</ul>
<h2 id="worked-example-protein-dna-with-introns">Worked Example: Protein-DNA with Introns<a class="headerlink" href="#worked-example-protein-dna-with-introns" title="Permanent link">&para;</a></h2>
<p>When aligning a protein sequence to genomic DNA containing introns:</p>
<ol>
<li><strong>DNA</strong> with introns: length <span class="arithmatex">\(\sim 3,000\)</span> to <span class="arithmatex">\(\sim 1,000,000\)</span> due to introns</li>
<li><strong>Protein</strong> sequence: length <span class="arithmatex">\(\sim\! \frac{|DNA|}{3}\)</span> but no introns</li>
</ol>
<p><strong>Strategy</strong>:
1. Identify or guess which reading frame(s) might be correct (try all frames 0, 1, 2, and possibly reverse complement)
2. Group DNA in blocks of 3 nucleotides (codons)
3. In the dynamic programming matrix, make the DNA index jump by 3 for each single step in the protein
4. Use a codon-to-amino-acid mapping in the dynamic programming's substitution step
5. Allow large gaps to account for introns, possibly using specialized "jumping alignment" techniques
6. If intron positions are unknown, incorporate a "splicing" model that allows skipping large segments in the DNA</p>
<h2 id="practical-applications">Practical Applications<a class="headerlink" href="#practical-applications" title="Permanent link">&para;</a></h2>
<p>Multiple sequence alignment is essential for:</p>
<ul>
<li><strong>Phylogenetic analysis</strong>: Reconstructing evolutionary relationships</li>
<li><strong>Structural prediction</strong>: Identifying conserved structural elements</li>
<li><strong>Functional annotation</strong>: Discovering functional domains and motifs</li>
<li><strong>Primer design</strong>: Creating primers that target conserved regions</li>
<li><strong>Metagenomics</strong>: Analyzing complex microbial communities</li>
</ul>
<h2 id="common-questions-and-answers">Common Questions and Answers<a class="headerlink" href="#common-questions-and-answers" title="Permanent link">&para;</a></h2>
<ol>
<li>
<p><strong>Q</strong>: Why is multiple sequence alignment (with sum-of-pairs scoring) NP-hard?<br />
<strong>A</strong>: Because the search space of all gap placements across <span class="arithmatex">\(D\)</span> sequences grows exponentially, and it has been formally proven that an optimal MSA (maximizing or minimizing certain standard scoring functions) is NP-hard.</p>
</li>
<li>
<p><strong>Q</strong>: How do codons affect DNA–protein alignment?<br />
<strong>A</strong>: DNA is read in triplets, so aligning each codon to one amino acid can improve accuracy. Gaps in the DNA must be multiples of 3 if they are truly coding segments, but introns complicate matters by allowing large "non-triplet" chunks to be skipped.</p>
</li>
<li>
<p><strong>Q</strong>: What if the sequences differ greatly in length?<br />
<strong>A</strong>: Large insertions/deletions or introns require special gap penalty regimes or specialized gene/spliced alignment methods. Standard linear gap scoring may overly penalize or distort the alignment.</p>
</li>
<li>
<p><strong>Q</strong>: How do we align 10 or more sequences in practice if it's NP-hard?<br />
<strong>A</strong>: Use heuristics such as progressive alignment guided by a phylogenetic or distance-based tree. Then refine the resulting alignment iteratively if needed.</p>
</li>
</ol>
<h2 id="practice-questions">Practice Questions<a class="headerlink" href="#practice-questions" title="Permanent link">&para;</a></h2>
<ol>
<li><strong>Pairwise vs. Multiple</strong>  </li>
<li>
<p>Define pairwise alignment and MSA. Why does dynamic programming suffice for the former but not scale to the latter?</p>
</li>
<li>
<p><strong>Gap Scoring</strong>  </p>
</li>
<li>
<p>Compare the difference between fixed gap penalties (<span class="arithmatex">\(\delta\)</span>) vs. affine gap costs (<span class="arithmatex">\(g + r \cdot L\)</span>), with <span class="arithmatex">\(g\)</span> = gap opening penalty, <span class="arithmatex">\(r\)</span> = extension penalty, <span class="arithmatex">\(L\)</span> = gap length. How do the DP recurrences change?</p>
</li>
<li>
<p><strong>Codon-Based Alignments</strong>  </p>
</li>
<li>
<p>Outline how you would adapt the Needleman–Wunsch (global alignment) recurrence if each step in DNA moves 3 nucleotides to match 1 amino acid. Where do you incorporate the codon → amino acid lookup?</p>
</li>
<li>
<p><strong>Introns and Large Gaps</strong>  </p>
</li>
<li>
<p>Suppose you want to align a cDNA of length 2,000 to a genomic DNA of length 50,000 (with introns). How might you modify or optimize standard DP to handle potentially huge gaps?</p>
</li>
<li>
<p><strong>Progressive Alignment</strong>  </p>
</li>
<li>Summarize the progressive alignment method. What is a "guide tree," and how do "profiles" help incorporate more than two sequences?</li>
</ol>
<h2 id="additional-reading">Additional Reading<a class="headerlink" href="#additional-reading" title="Permanent link">&para;</a></h2>
<ul>
<li><strong>Mäkinen et al. (2015)</strong>: Chapter 6, "Alignments," especially:</li>
<li>§6.4 (Biological sequence alignment: global, local, overlap)</li>
<li>§6.5 (Gene alignment and splicing)</li>
<li>§6.6 (Multiple alignment, hardness, progressive MSA)</li>
<li><strong>PI-AVG-01-Multiple_Sequence_Alignment.pdf</strong>:</li>
<li>Slides on pairwise alignment definitions, edit distance, dynamic programming</li>
<li>Sections on local, semiglobal, overlap alignments</li>
<li>Material on heuristic algorithms for MSA (progressive/branch-and-bound)</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.sections", "navigation.top", "search.highlight", "content.action.view", "content.code.copy"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8b220af.min.js"></script>
      
        <script src="../../assets/_markdown_exec_pyodide.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
        <script src="../../javascripts/mathjax.js"></script>
      
    
  </body>
</html>